README
NAME
    CGI::Application::Plugin::DevPopup - Runtime cgiapp info in a popup
    window

VERSION
    Version 1.01

SYNOPSIS
  End user information
    This module provides a plugin framework for displaying runtime
    information about your CGI::Application app in a popup window. A sample
    Timing plugin is provided to show how it works:

        BEGIN { $ENV{'CAP_DEVPOPUP_EXEC'} = 1; } # turn it on for real
        use CGI::Application::Plugin::DevPopup;
        use CGI::Application::Plugin::DevPopup::Timing;

        The rest of your application follows
        ...

    Now whenever you access a runmode, a window pops up over your content,
    showing information about how long the various stages have taken. Adding
    other CAP::DevPopup plugins will get you more information. A HTML::Tidy
    plugin showing you how your document conforms to W3C standards is
    available: see CGI::Application::Plugin::HtmlTidy.

    The output consists of a Table of Contents, and a bunch of reports. A
    rough translation into plain text could look like this:

        Devpopup report for My::App -> add_timing

        * Timings - Total runtime: 3.1178 sec.

        +-----------------------------------------------------------------------+
        | Timings                                                               |
        +-----------------------------------------------------------------------+
        | Application started at: Thu Sep 22 02:55:35 2005                      |
        | From                       To                         Time taken      |
        |-----------------------------------------------------------------------|
        | init                       prerun                     0.107513 sec.   |
        | prerun                     before expensive operation 0.000371 sec.   |
        | before expensive operation after expensive operation  3.006688 sec.   |
        | after expensive operation  load_tmpl(dp.html)         0.000379 sec.   |
        | load_tmpl(dp.html)         postrun                    0.002849 sec.   |
        +-----------------------------------------------------------------------+

    The reports expand and collapse by clicking on the ToC entry or the
    report header.

    You can see a (developer) version in action here:
    <http://oss.rhesa.com/cpan/cap-devpopup/example>.

  Developer information
    Creating a new plugin for DevPopup is fairly simple. CAP::DevPopup
    registers a new callback point (named "devpopup_report"), which it uses
    to collect output from your plugin. You can add a callback to that
    point, and return your formatted output from there. The callback has
    this signature:

        sub callback($cgiapp_class, $outputref)

    You pass your output to the devpopup object by calling

        $cgiapp_class->devpopup->add_report(
                    title   => $title,
                    summary => $summary,
                    report  => $body,
        );

    You are receiving $outputref, because DevPopup wants to be the last one
    to be called in the postrun callback. If you had wanted to act at
    postrun time, then please do so with this variable, and not through a
    callback at postrun.

  The "on" switch
    Since this is primarily a development plugin, and you wouldn't want it
    to run in your production code, an environment variable named
    CAP_DEVPOPUP_EXEC has to be set to 1 for this module to function, and it
    must be present at compile time. This means you should place it in a
    BEGIN{} block, or use SetEnv or PerlSetEnv (remember to set those before
    any PerlRequire or PerlModule lines).

    Absense of the environment variable turns this module into a no-op:
    while the plugin and its plugins are still loaded, they won't modify
    your output.

Available Plugins
    o CGI::Application::Plugin::DevPopup::Timing,
    CGI::Application::Plugin::DevPopup::Log and
    CGI::Application::Plugin::DevPopup::HTTPHeaders are bundled with this
    distribution.
    o CGI::Application::Plugin::HtmlTidy integrates with this module.
    o CGI::Application::Plugin::TT integrates with this module.

EXPORTS
    * devpopup
        This method is the only one exported into your module, and can be
        used to access the underlying DevPopup object. See below for the
        methods that this object exposes.

METHODS
    * add_report( %fields )
        Adds a new report about the current run of the application. The
        following fields are supported:

        * title A short title for your report

        * summary
                An optional one- or two-line summary of your findings

        * report
                Your full output

        * severity
                An optional value specifying the importance of your report.
                Accepted values are qw/debug info warning error fatal/. This
                value is used to color-code the report headers.

        * script
                If you have custom javascript, then please pass it in
                through this field. Otherwise if it's embedded in your
                report, it will break the popup window. I will take care of
                the surrounding "<script"> tags, so just the code body is
                needed.

INSTALLATION
    INSTALLATION

    To install this module, run:

        cpan CGI::Application::Plugin::DevPopup

    To mnually install this module, run the following commands:

        perl Makefile.PL
        make
        make test
        make install

SEE ALSO
    CGI::Application. CGI::Application::Plugin::DevPopup::Timing

AUTHOR
    Rhesa Rozendaal, rhesa@cpan.org

BUGS
    Please report any bugs or feature requests to
    bug-cgi-application-plugin-devpopup@rt.cpan.org, or through the web
    interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=CGI-Application-Plugin-D
    evPopup>. I will be notified, and then you'll automatically be notified
    of progress on your bug as I make changes.

ACKNOWLEDGEMENTS
    Mark Stosberg for the initial idea, and for pushing me to write it.
    Sam Tregar for providing me with the skeleton cgiapp_postrun.
    Joshua Miller for providing the ::Log plugin.
    Everybody on the cgiapp mailinglist and on #cgiapp for cheering me on
    :-)

COPYRIGHT & LICENSE
    Copyright 2005-2007 Rhesa Rozendaal, all rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for CGI-Application-Plugin-DevPopup

1.07    2011-07-17  Rhesa Rozendaal <rhesa@cpan.org>
    - Document HTTPHeaders output

1.06    2010-01-10  Rhesa Rozendaal <rhesa@cpan.org>
    - Better XHTML (Alexandr Ciornii)

1.05    2010-01-07  Rhesa Rozendaal <rhesa@cpan.org>
    - Added repository to META.yml. No code changes.

1.04    2009-12-25  Rhesa Rozendaal <rhesa@cpan.org>
    - RT #53100 (Serguei Trouchelle)

1.03    2009-10-20  Rhesa Rozendaal <rhesa@cpan.org>
    - RT #50096 (Serguei Trouchelle)
    - RT #42315 (Lyle Hopkins)

1.02    2009-01-15  Rhesa Rozendaal <rhesa@cpan.org>
    - Moved to Dist::Zilla
    - Migrated repository to github

1.01    2008-10-24  Rhesa Rozendaal <rhesa@cpan.org>
    - RT #35088
    - CPANTS fixes

1.00    2007-09-27  Rhesa Rozendaal <rhesa@cpan.org>
    - added ::Log plugin, kindly provided by Joshua Miller

0.951   2006-11-20  Rhesa Rozendaal <rhesa@cpan.org>
    - pod fixes
    - dist improvements conforming to CPANTS

0.95    2005-12-07  Rhesa Rozendaal <rhesa@cpan.org>
    - added "severity" attribute to reports.
    - cosmetic changes in css and html
    - added log report (from Log::Dispatch or STDERR)

0.94    2005-11-23  Rhesa Rozendaal <rhesa@cpan.org>
    - scrollbars didn't show up after all (reported by Robert Lawson)

0.93    2005-10-05  Rhesa Rozendaal <rhesa@cpan.org>
    - Made popup resizable and scrollable in IE.

0.92    2005-10-05  Rhesa Rozendaal <rhesa@cpan.org>
    - Moved HTTP Headers to its own module

0.91    2005-10-04  Rhesa Rozendaal <rhesa@cpan.org>
    - Added HTTP Headers report
    - Updated docs to reflect availability of CAP::HtmlTidy

0.02    2005-09-22  Rhesa Rozendaal <rhesa@cpan.org>
    - Added master switch in the form of $ENV{CAP_DEVPOPUP_EXEC}
        (thanks to markstos and hide on #cgiapp)

    - Added a hook to ::Timing at devpopup_addtiming for your own timing needs
    - Added name of template in load_tmpl timing point
        (thanks to Michael Graham on the mailing list)

    - Cosmetic changes in html

0.01    2005-09-21  Rhesa Rozendaal <rhesa@cpan.org>
    - Initial revision, basic html and a simple Timing plugin.



lib/CGI/Application/Plugin/DevPopup/Timing.pm
package CGI::Application::Plugin::DevPopup::Timing;

use strict;
use base qw/Exporter/;
use Time::HiRes qw/gettimeofday tv_interval/;
my $start = [gettimeofday];

sub import
{
    my $c = scalar caller;
    $c->add_callback( 'devpopup_report', \&_timer_report );
    $c->new_hook('devpopup_addtiming');
    $c->add_callback( 'devpopup_addtiming', \&_add_time );
    foreach my $stage (qw/ init prerun load_tmpl /)
    {
        $c->add_callback( $stage, sub { _add_time( shift(), $stage, @_ ) } );
    }
    goto &Exporter::import;
}

sub _timer_report
{
    my $app = shift;

    my $self = _new_or_self($app);
    unshift @$self, { dec => 'start', tod => $start };
    _add_time( $app, 'postrun' );
    $app->devpopup->add_report(
        title   => 'Timings',
        summary => 'Total runtime: ' . tv_interval( $self->[1]{tod}, $self->[-1]{tod} ) . ' sec.',
        report  => '<style>
            th { text-align:left; border-bottom:solid 1px black; }
            </style>' .
            'Application started at: ' . scalar( gmtime( $start->[0] ) ) . ' GMT<br/>' .
            '<table width="100%"><tr><th>From</th><th>To</th><th>Time taken</th></tr>' .
            join(
              $/,
              map {
                my $time = tv_interval( $self->[$_]{tod}, $self->[ $_ + 1 ]{tod} );
                "<tr><td>$self->[$_]{desc}</td><td>$self->[ $_ + 1 ]{desc}</td><td>$time sec.</td></tr>"
              } ( 1 .. $#$self - 1 )
            )
            . '</table>'
    );
}

sub _add_time
{
    my $app   = shift;
    my $stage = shift;
    $stage .= '(' . $_[-1] . ')' if $stage =~ /load_tmpl/;

    my $self = _new_or_self($app);
    push @$self, { desc => $stage, tod => [gettimeofday] };
}

sub _new_or_self
{
    my $app  = shift;
    my $self = $app->param('__CAP_DEVPOPUP_TIMER');
    unless ($self)
    {
        $self = bless [], __PACKAGE__;
        $app->param( '__CAP_DEVPOPUP_TIMER' => $self );
    }
    return $self;
}

1;    # End of CGI::Application::Plugin::DevPopup::Timing

__END__

=head1 NAME

CGI::Application::Plugin::DevPopup::Timing - show timing information about cgiapp stages

=head1 SYNOPSIS

    use CGI::Application::Plugin::DevPopup;
    use CGI::Application::Plugin::DevPopup::Timing;

    The rest of your application follows
    ...

Output will look roughly like this:

    Timings
    Total runtime: 3.1178 sec.
    Application started at: Thu Sep 22 02:55:35 2005

    From                        To                              Time taken
    -------------------------------------------------------------------------
    init                        prerun                          0.107513 sec.
    prerun                      before expensive operation      0.000371 sec.
    before expensive operation  after expensive operation       3.006688 sec.
    after expensive operation   load_tmpl(dp.html)              0.000379 sec.
    load_tmpl(dp.html)          postrun                         0.002849 sec.

=head1 ADD_TIMING

You can add your own timing points within your application by calling the hook at C<devpopup_addtiming> with a short label. The table above was created with the following code:

    $self->call_hook('devpopup_addtiming', 'before expensive operation');
    sleep 3;
    $self->call_hook('devpopup_addtiming', 'after expensive operation');

=head1 SEE ALSO

L<CGI::Application::Plugin::DevPopup>, L<CGI::Application>

=head1 AUTHOR

Rhesa Rozendaal, L<rhesa@cpan.org>

=head1 BUGS

Please report any bugs or feature requests to
L<bug-cgi-application-plugin-devpopup@rt.cpan.org>, or through the web
interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=CGI-Application-Plugin-DevPopup>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=head1 COPYRIGHT & LICENSE

Copyright 2005 Rhesa Rozendaal, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

